### PR TITLE
Use tinygemm bfloat16 / int4 kernel whenever possible

### PR DESCRIPTION
**Pull request by dacorvo**
# What does this PR do?

This adds a new `QBitsTensor` type, `TinyGemmQBitsTensor`, that formats data, scale and shift as expected by the `tinygemm` kernels included in pytorch (`torch._weight_int8pack_mm`).

This greatly speeds up the inference with `int4` weights on CUDA and CPU for `bfloat16` models (almost on par with `bfloat16` weights).

Note that `tinygemm` kernels are also available for MPS, but only starting from pytorch 2.4.

### CODE REVIEW

File: test/tensor/qbits/test_awq_qbits_tensor.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    # Verify the dequantized tensors are identical
+    assert torch.equal(awqbt.dequantize(), qbt.dequantize())
+    # Now verify that we can reconstruct the QBitsTensor
+    new_qbt = awqbt.qbits_tensor()
+    assert new_qbt.dtype == dtype
+    assert new_qbt.qtype == qtype
+    assert new_qbt.shape == shape
+    assert torch.equal(new_qbt._data, qbt._data)
+    assert torch.equal(new_qbt._scale, qbt._scale)
+    assert torch.equal(new_qbt._shift, qbt._shift)
```
Comment (dacorvo) [inline test/tensor/qbits/test_awq_qbits_tensor.py:56]: There should be a move test here also.

File: optimum/quanto/tensor/qweight.py [hunk 2]
```diff
@@ -42,6 +47,9 @@ def quantize_weight(
         group_size (`Optional[int]`): The quantization group size
         optimizer (`Optional[quanto.Optimizer]`): An optimizer to evaluate the scale if not provided.
             Defaults to a max Optimizer.
+        zeropoint (`bool`): Allow an exact representation of zero. If True, the shifts are stored as
+            integer instead of float, which results in a slightly smaller model, but might also reduce
+            the model performance. Defaults to True.
```
Comment (dacorvo) [inline optimum/quanto/tensor/qweight.py:52]:
```suggestion
            the model performance. Defaults to False.
```

**Comment by SunMarc** [comment]
> This greatly speeds up the inference with int4 weights on CUDA and CPU for bfloat16 models (almost on par with bfloat16 weights).

Wow ! How does it compare to awq kernels ? Nice to see that they have added support to mps kernels too. It would be great to do a table in order to track which kernels are used for each backend !

**Comment by dacorvo** [comment]
> Wow ! How does it compare to awq kernels ? 
It looks almost as fast, at least on the generation benchmark. It should be tested with different batch sizes, input length to be sure.
>It would be great to do a table in order to track which kernels are used for each backend ! 
I was thinking exactly the same thing. Although it is a bit complicated since some kernels are only available for some dtypes, so maybe having pairs of (input, weight) types in rows, and backends in columns ... Feel free to submit something !

**Pull request merged**